Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -1429,6 +1463,222 @@ TEST_F(SerializerTest, EncodesResumeTokens) {
   ExpectRoundTrip(model, proto);
 }
 
+TEST_F(SerializerTest, EncodesListenRequestLabels) {
+  core::Query q = Query("docs");
+
+  std::map<QueryPurpose, std::unordered_map<std::string, std::string>>
+      purpose_to_label = {
+          {QueryPurpose::Listen, {}},
+          {QueryPurpose::LimboResolution,
+           {{"goog-listen-tags", "limbo-document"}}},
+          {QueryPurpose::ExistenceFilterMismatch,
+           {{"goog-listen-tags", "existence-filter-mismatch"}}},
+      };
+
+  for (const auto& p : purpose_to_label) {
+    QueryData model(q, 1, 0, p.first);
+
+    auto result = serializer.EncodeListenRequestLabels(model);
+    EXPECT_EQ(result, p.second);
+  }
+}
+
+TEST_F(SerializerTest, DecodesMutationResult) {
+  std::vector<FieldValue> transformations({FieldValue::FromBoolean(true),
+                                           FieldValue::FromInteger(1234),
+                                           FieldValue::FromString("string")});
+  auto version = Version(123456789);
+  MutationResult model(version, std::move(transformations));
+
+  v1::WriteResult proto;
+
+  proto.mutable_update_time()->set_seconds(version.timestamp().seconds());
+  proto.mutable_update_time()->set_nanos(version.timestamp().nanoseconds());
+  proto.mutable_transform_results()->Add();
+  (*proto.mutable_transform_results())[0] = ValueProto(true);
+  proto.mutable_transform_results()->Add();
+  (*proto.mutable_transform_results())[1] = ValueProto(1234);
+  proto.mutable_transform_results()->Add();
+  (*proto.mutable_transform_results())[2] = ValueProto("string");
+
+  SCOPED_TRACE("DecodesMutationResult");
+  ExpectDeserializationRoundTrip(model, proto, Version(10000000));
+}
+
+TEST_F(SerializerTest, DecodesMutationResultWithNoUpdateTime) {
+  MutationResult model(Version(10000000), {});
+
+  v1::WriteResult proto;
+
+  SCOPED_TRACE("DecodesMutationResultWithNoUpdateTime");
+  ExpectDeserializationRoundTrip(model, proto, Version(10000000));
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithAddedTargetChange) {
+  WatchTargetChange model(WatchTargetChangeState::Added, {1, 2},
+                          ByteString("resume_token"));
+
+  v1::ListenResponse proto;
+
+  proto.mutable_target_change()->set_target_change_type(
+      v1::TargetChange_TargetChangeType::TargetChange_TargetChangeType_ADD);
+  proto.mutable_target_change()->add_target_ids(1);
+  proto.mutable_target_change()->add_target_ids(2);
+  proto.mutable_target_change()->set_resume_token("resume_token");
+
+  SCOPED_TRACE("DecodesListenResponseWithAddedTargetChange");
+  ExpectDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithRemovedTargetChange) {
+  WatchTargetChange model(WatchTargetChangeState::Removed, {1, 2},
+                          ByteString("resume_token"),
+                          Status{Error::PermissionDenied, "Error message"});
+
+  v1::ListenResponse proto;
+
+  proto.mutable_target_change()->set_target_change_type(
```

Comment:
Unless you've been given advice to the contrary, you might consider adding temporaries to make it clearer what each line is doing:

```c++
auto change = proto.mutable_target_change();
change->add_target_ids(1);
// ...
```

Reply:
Done.